### ISSUE TITLE
dbReadTable as_of argument fails when table schema has changed

### ISSUE THREAD

**Issue by n8layman** [ecohealthalliance/doltr#46]
as_of argument works as expected unless schema has changed w.r.t. HEAD commit. For example after adding a table column. dbReadTable then produces the following error when accessing a commit before the change occurred:

`Error: The number of parameters in bound buffers differs from number of columns in resultset [2057]`

This feels like mysql or dolt error and not doltr. Recording here in case it happens to others. 

OSX: 12.4
Arch: arm64
R: 4.2.0
RMariaDB: 1.2.1
dolt: 0.40.0
mysql: Ver 8.0.29 for macos12.2 on arm64 (Homebrew)

**Comment by noamross** [comment]
Feel free to ping the dolt Discord or file an issue if you're able to create an MRE.  Here's how I filed an issue for a dolt error that included the dolt server logs:  https://github.com/dolthub/dolt/issues/3418

**Comment by n8layman** [comment]
Will do later this evening. I want to test on another machine to see if I can reproduce the issue first. Thanks!

**Comment by timsehn** [comment]
It's not an obvious bug on our side.
```
test-as-of $ dolt init
Successfully initialized dolt data repository.
test-as-of $ dolt sql -q "create table test (pk int, c1 int, primary key(pk))" 
test-as-of $ dolt sql -q "insert into test values (0,0), (1,1)"
Query OK, 2 rows affected
test-as-of $ dolt sql -q "select * from test as of 'HEAD'" 
error on line 1 for query select * from test as of 'HEAD': table not found: test
table not found: test
test-as-of $ dolt sql -q "select * from test" 
+----+----+
| pk | c1 |
+----+----+
| 0  | 0  |
| 1  | 1  |
+----+----+
test-as-of $ dolt commit -am "Table and some values"
commit c0mkp3dic4d2hb1vbeam9ivt50e11plp (HEAD -> main) 
Author: Tim Sehn <tim@dolthub.com>
Date:  Thu Jun 09 13:26:00 -0700 2022

        Table and some values

test-as-of $ dolt sql -q "select * from test as of 'HEAD'" 
+----+----+
| pk | c1 |
+----+----+
| 0  | 0  |
| 1  | 1  |
+----+----+
test-as-of $ dolt sql -q "select * from test as of 'c0mkp3dic4d2hb1vbeam9ivt50e11plp'" 
+----+----+
| pk | c1 |
+----+----+
| 0  | 0  |
| 1  | 1  |
+----+----+
test-as-of $ dolt sql -q "select * from test as of 'HEAD^'" 
error on line 1 for query select * from test as of 'HEAD^': table not found: test
table not found: test
test-as-of $ dolt sql -q "alter table test add column c2 int"
test-as-of $ dolt commit -am "Added column c2"
commit juo78dcl3fl9t6pskmu2bn6jlfi9babv (HEAD -> main) 
Author: Tim Sehn <tim@dolthub.com>
Date:  Thu Jun 09 13:26:56 -0700 2022

        Added column c2

test-as-of $ dolt sql -q "select * from test as of 'HEAD'" 
+----+----+------+
| pk | c1 | c2   |
+----+----+------+
| 0  | 0  | NULL |
| 1  | 1  | NULL |
+----+----+------+
test-as-of $ dolt sql -q "select * from test as of 'HEAD^'" 
+----+----+
| pk | c1 |
+----+----+
| 0  | 0  |
| 1  | 1  |
+----+----+
```

I'll look where that error is in our code and see if we can see how a schema change + as of could trigger it.

**Comment by timsehn** [comment]
This may be a client issue where we're not setting the correct dimensions of the return type. So, in my local case, it prints fine but a client expects a different number of columns than are returned. Testing that.

**Comment by timsehn** [comment]
@noamross You all are using the MariaDB client right? `mysql` client seems to be fine with what we're doing.

```
dolt $ mysql --host 127.0.0.1 --port 3306 -uroot
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 1
Server version: 5.7.9-Vitess 

Copyright (c) 2000, 2022, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| test_as_of         |
+--------------------+
2 rows in set (0.00 sec)

mysql> use test_as_of;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
mysql> select * from test as of 'HEAD^'
    -> ;
+------+------+
| pk   | c1   |
+------+------+
|    0 |    0 |
|    1 |    1 |
+------+------+
2 rows in set (0.00 sec)
```

I'll try the MariaDB client.

**Comment by n8layman** [comment]
Yeah we are. Let's move this to an issue on your repo though before it gets too big! I'll create it now.

**Comment by timsehn** [comment]
Thanks!

**Comment by n8layman** [comment]
Moving to dolt issue [#3583](https://github.com/dolthub/dolt/issues/3583)

**Comment by n8layman** [comment]
Issue appears resolved in dolt release 0.40.9.
